enclave/src/polkadex_gateway.rs

```diff
@@ -0,0 +1,137 @@
+use log::*;
+use polkadex_sgx_primitives::AccountId;
+use polkadex_sgx_primitives::types::{Order, OrderSide, OrderType, OrderUUID, TradeEvent};
+use sgx_types::{sgx_status_t, SgxResult};
+
+use crate::polkadex;
+use crate::polkadex_balance_storage;
+use crate::polkadex_orderbook_storage;
+
+/// Place order function does the following
+/// 1. authenticate
+/// 2. mutate balances (reserve amount offered in order)
+/// 3. store_order (async)
+/// 4. send order to OpenFinex API
+/// 5. report OpenFinex API result to sender
+pub fn place_order(main_account: AccountId, proxy_acc: Option<AccountId>, order: Order) -> SgxResult<()> {
+    // Authentication
+    authenticate_user(main_account.clone(), proxy_acc)?;
+    // Mutate Balances
+    match order.order_type {
+        OrderType::LIMIT => {
+            if let Some(price) = order.price {
+                match order.side {
+                    OrderSide::BID => {
+                        // TODO: Is it safe to use saturating_mul here?
+                        let amount = price.saturating_mul(order.quantity);
+                        polkadex_balance_storage::reserve_balance(&main_account, order.market_id.quote, amount)?;
+                    }
+                    OrderSide::ASK => {
+                        // TODO: Is it safe to use saturating_mul here?
+                        polkadex_balance_storage::reserve_balance(&main_account, order.market_id.base, order.quantity)?;
+                    }
+                }
+            } else {
+                error!("Price not given for a limit order");
+                return Err(sgx_status_t::SGX_ERROR_UNEXPECTED); // TODO: It would be better if we can return a error response for RPC to handle
```

Comment:
SgxResult is only necessary when returning from an ecall or ocall. Because function will be called by the RPC API Felix is implementing, which is not an ecall, no SgxResult return value is necessary. So you can define any error you want and return it with JSON RPC to the client. Substrate mostly works with enums, but simple Strings or Structs are also possible. Depends on what you want to return..?

Reply:
Okay got it